tracer/slog.go

```diff
@@ -0,0 +1,158 @@
+package tracer
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"log/slog"
+	"strings"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgconn"
+)
+
+type LogTracer struct {
+	Logger           *slog.Logger
+	LogAllQueries    bool
+	LogFailedQueries bool
+	// replace placeholders with arguments useful for local debugging
+	LogValues bool
+	// enabled if non-zero value is provided
+	LogSlowQueriesThreshold time.Duration
+
+	// give client power to change each section which is being logged
+	StartQueryHook  func(ctx context.Context, query string, args []any)
+	SlowQueryHook   func(ctx context.Context, query string, duration time.Duration)
+	EndQueryHook    func(ctx context.Context, query string, duration time.Duration)
+	FailedQueryHook func(ctx context.Context, query string, err error)
+}
+
+func NewLogTracer(logger *slog.Logger, opts ...Option) *LogTracer {
+	logStart := func(ctx context.Context, query string, args []any) {
+		if logger != nil {
+			logger.LogAttrs(ctx, slog.LevelInfo, "query start", slog.String("query", query), slog.Any("args", args))
+		}
+	}
+
+	logSlowQuery := func(ctx context.Context, query string, duration time.Duration) {
+		if logger != nil {
+			logger.LogAttrs(ctx, slog.LevelWarn, "slow query took", slog.Any("query", query), slog.Duration("duration", duration))
+		}
+	}
+
+	logEnd := func(ctx context.Context, query string, duration time.Duration) {
+		if logger != nil {
+			logger.LogAttrs(ctx, slog.LevelInfo, "query end", slog.Any("query", query), slog.Duration("duration", duration))
+		}
+	}
+
+	logFailed := func(ctx context.Context, query string, err error) {
+		if logger != nil {
+			logger.LogAttrs(ctx, slog.LevelError, "query failed", slog.Any("query", query), slog.String("err", err.Error()))
+		}
+	}
+
+	cfg := &config{
+		logAllQueries:           false,
+		logFailedQueries:        false,
+		logValues:               false,
+		logSlowQueriesThreshold: 0,
+		logStartHook:            logStart,
+		logSlowQueryHook:        logSlowQuery,
+		logEndQueryHook:         logEnd,
+		logFailedQueryHook:      logFailed,
+	}
+
+	for _, opt := range opts {
+		opt.apply(cfg)
+	}
+
+	return &LogTracer{
+		LogAllQueries:           cfg.logAllQueries,
+		LogFailedQueries:        cfg.logFailedQueries,
+		LogValues:               cfg.logValues,
+		LogSlowQueriesThreshold: cfg.logSlowQueriesThreshold,
+		StartQueryHook:          cfg.logStartHook,
+		SlowQueryHook:           cfg.logSlowQueryHook,
+		EndQueryHook:            cfg.logEndQueryHook,
+		FailedQueryHook:         cfg.logFailedQueryHook,
+	}
+}
+
+func (l *LogTracer) TraceQueryStart(ctx context.Context, _ *pgx.Conn, data pgx.TraceQueryStartData) context.Context {
+	query := data.SQL
+	if l.LogValues {
+		for i, placeholder := range data.Args {
+			query = strings.Replace(query, fmt.Sprintf("$%d", i+1), fmt.Sprintf("%v", placeholder), 1)
```

Comment:
This might be problematic because `query` is scanned whole every time, and the final size depends on the number of arguments. Also, the data that is left in place after replacing it is going to be scanned too. I am thinking about a few scenarios where this might not be desirable especially because we (or I, at least) tend to copy and paste logs without reading them in whole, for instance:

Extremely large SQL statement:

```
`INSERT INTO images (binary_data) VALUES($1)`, []byte(very-long-blob-of-unreadable-binary-data...)
```

Arguments containing the replacement pattern itself:

```
`INSERT INTO accounts (names) VALUES ($1, $2, $3)`, "$2", "$2", "$4");
```

Specially crafted bogus logs:

```
`INSERT INTO accounts (names) VALUES ($1)`, "'Robert'); DROP TABLE Students;--"
```

or

```
`INSERT INTO accounts (names) VALUES ($1)`, "NULL) ...lots of whitespace and then something harmful that could be executed after copy-pasting the log somewhere..."
```
My point is that this opens the door to manipulating logs via their arguments. If we want to do in-place replacement of the arguments, I suggest byte-scanning the string for `$1`, `$2`, ... `$n` and keeping a cursor with the current position instead of search-replacing the whole string on each iteration. Also, `%q` might be safer than `%v` in this particular case (assuming the goal isn't creating valid SQL, in which case neither `%q` nor `%v` would help).

Reply:
Like this: https://github.com/goware/pgkit/pull/17/commits/132de3e3de69dc6fe00117c93bc7eea21c7d9651 ?

I mean this feature was only meant for local debugging if needed, because we can't log arguments on prod because of GDPR and it would not be good idea overall I guess. 

Anyway you have a good point that sometimes the logs might be too much because we work with binary data a lot in our app and it might be unusable in the end.

Reply:
That'll do the trick, thanks!